### ISSUE TITLE
ggml-org (ggml)

### ISSUE THREAD

**Issue by TAIM-Tuerkiye** [ggml-org/ggml#1481]
Topic: ggml_build_backward_expand crashes in ggml-opt on ROCm (and Vulkan) 

Summary 
ggml_build_backward_expand raises a segfault / abort when called via ggml_opt_alloc during LoRA training
finetune. The crash occurs regardless of GPU backend (HIP, Vulkan, CPU) and appears to be in the backward graph
construction itself, not in a specific GPU kernel. 
System 
OS: Kubuntu 24.04, Linux 6.17.0-23-generic
GPU: AMD Radeon RX 9070 XT (RDNA4, 16GB VRAM)
ROCm: 7.2
ggml: commit b73e75a from QVAC repository (latest as of 2026-05-09)
Build: CMake with -DGGML_HIP=ON (also tested with -DGGML_VULKAN=ON) 
Reproduction 
# Build
cmake -B build -DGGML_HIP=ON -DLLAMA_CURL=OFF -DCMAKE_BUILD_TYPE=Release
cmake --build build --target llama-finetune-lora -j
 # Run (tested with TinyLlama 1.1B GGUF)
./build/bin/llama-finetune-lora \
--model /path/to/tinyllama.gguf \
-f /tmp/test_data.txt \
--output-adapter /tmp/lora.gguf \
--learning-rate 1e-4 \
--num-epochs 1 \
--batch-size 2 \
--ctx-size 256 \
--lora-rank 4

Backtrace 
#0 __GI___wait4 () at ../sysdeps/unix/sysv/linux/wait4.c:30
#1 ggml_print_backtrace () from libggml-base.so.0
#2 ggml_abort () from libggml-base.so.0
#3 ggml_build_backward_expand () from libggml-base.so.0
#4 ggml_opt_build (ggml_opt_context*) () from libggml-base.so.0
#5 ggml_opt_alloc () from libggml-base.so.0
#6 llama_context::opt_epoch_iter () from libllama.so.0
#7 llama_context::opt_epoch () from libllama.so.0
#8 main ()

Analysis 
ggml_build_backward_expand at frame #3 is where the abort happens
This is called during graph construction for the backward pass (before any GPU kernel launch)
Happens on CPU, HIP, and Vulkan backends — confirming it’s not backend-specific
Likely an unhandled operation type or a graph construction invariant violation
The crash happens later in the backward pass, after the dataset init (which appears to be a separate issue, seeQVAC report) 

Expected behavior 
Backward pass graph should be constructed gracefully for standard transformer layer operations (forward +
backward). If certain ops are unsupported, a clear error message should indicate which op is missing a backward
implementation.

**Comment by Kaihui-AMD** [comment]
Hi @TAIM-Tuerkiye ,

This isn't ROCm-specific. Your analysis is spot on the crash is in `ggml_build_backward_expand`, which is CPU-side graph construction before any GPU backend is involved. Your cross-backend testing confirms this.

Known issue in ggml's backward pass. The same assertion failure is reported in [llama.cpp#22479](https://github.com/ggml-org/llama.cpp/issues/22479) (open, same backtrace), and [llama.cpp#21037](https://github.com/ggml-org/llama.cpp/issues/21037) documents 5 cascading bugs in the backward pass including the missing `SET_ROWS` and `FLASH_ATTN_EXT` gradient handling. These are llama.cpp issues but the backward pass code lives in the shared `ggml.c`, which is why they're relevant here.

For LoRA training on AMD GPUs right now, [HuggingFace PEFT with PyTorch ROCm](https://rocm.docs.amd.com/en/latest/how-to/rocm-for-ai/fine-tuning/single-gpu-fine-tuning-and-inference.html#fine-tuning) is probably the most reliable path, though the primary testing is on Instinct GPUs rather than consumer RDNA4.

**Comment by TAIM-Tuerkiye** [comment]
That's the workflow we have chosen and it works with 273 tokens/s!
